pkg/util/watch/metrics.go

```diff
@@ -0,0 +1,180 @@
+// - failing_current (1 iff the watcher's last request failed *and* it's waiting to retry, else 0)
+//
+// Prefixes are typically of the form "COMPONENT_watchers_" (e.g. "autoscaling_agent_watchers_").
+// Separate reporting per call to Watch is automatically done with the "watcher_instance" label
+// attached to the metrics, using MetricsConfig.
+//
+// A brief note about "alive" and "failing": Reading from a pair of collectors is fundamentally
+// racy. It may be possible to temporarily view "failing" but not "alive".
+type Metrics struct {
+	clientCallsTotal    *prometheus.CounterVec
+	relistRequestsTotal *prometheus.CounterVec
+	eventsTotal         *prometheus.CounterVec
+	errorsTotal         *prometheus.CounterVec
+	aliveCurrent        *prometheus.GaugeVec
+	failingCurrent      *prometheus.GaugeVec
+}
+
+type MetricsConfig struct {
+	Metrics
+	// Instance provides the value of the "watcher_instance" label that will be applied to all
+	// metrics collected for the Watch call
+	Instance string
+}
+
+const metricInstanceLabel = "watcher_instance"
+
+// NewMetrics creates a new set of metrics for one or many Watch calls
+//
+// All metrics' names will be prefixed with the provided string.
+func NewMetrics(prefix string) Metrics {
+	return Metrics{
+		clientCallsTotal: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: fmt.Sprint(prefix, "client_calls_total"),
+				Help: "Number of calls to k8s client.{Watch,List}, labeled by method",
+			},
+			[]string{metricInstanceLabel, "method"},
+		),
+		relistRequestsTotal: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: fmt.Sprint(prefix, "relist_requests_total"),
+				Help: "Number of internal manual relisting requests",
+			},
+			[]string{metricInstanceLabel},
+		),
+		eventsTotal: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: fmt.Sprint(prefix, "events_total"),
+				Help: "Number of k8s watch.Events that have occurred, including errors, labeled by type",
+			},
+			[]string{metricInstanceLabel, "type"},
+		),
+		errorsTotal: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: fmt.Sprint(prefix, "errors_total"),
+				Help: "Number of errors, either error events or re-list errors, labeled by source",
+			},
+			[]string{metricInstanceLabel, "source"},
+		),
+		aliveCurrent: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: fmt.Sprint(prefix, "alive_current"),
+				Help: "For each watcher, 1 iff the watcher is currently running or failing, else 0",
+			},
+			[]string{metricInstanceLabel},
+		),
+		failingCurrent: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: fmt.Sprint(prefix, "failing_current"),
+				Help: "For each watcher, 1 iff the watcher's last request failed *and* it's waiting to retry, else 0",
+			},
+			[]string{metricInstanceLabel},
+		),
+	}
+}
+
+// MustRegister registers all the collectors in the Metrics
+func (m *Metrics) MustRegister(reg *prometheus.Registry) {
```

Comment:
why keep this seperate and not use `util.RegisterMetric` for these?

Reply:
Record of discussion: reasoning was that `watch.Metrics` wants to be purely internal to the `watch` package, so we can't really directly expose the collector(s).

Reply:
Is there a case where you would construct a `*Metrics` object without registering metrics? It doesn't look like that's the case, and you could just pass a registry to the `NewMetrics` constructor, am I wrong about this?

Reply:
> Is there a case where you would construct a `*Metrics` object without registering metrics?

That's currently the case in the autoscaler-agent. Setting up the scheduler store:

https://github.com/neondatabase/autoscaling/blob/9ee46220f1c0dba14bf576f0f0bbde20311c0f88/pkg/agent/entrypoint.go#L56-L61

uses a `watch.Metrics`, and is required before calling `r.newAgentState`:

https://github.com/neondatabase/autoscaling/blob/9ee46220f1c0dba14bf576f0f0bbde20311c0f88/pkg/agent/entrypoint.go#L70

which produces the registry.

---

All that aside, I think it's generally preferable to allow constructing a `watch.Metrics` without registering it — if only for the possibility of future testing.

Reply:
I think, in general, it's better to prevent a pretty easy to make mistake (creating a new metric without registering it) rather than making the code a bit (though, to be fair, only a bit) harder to maintain, to support theoretical future testing (particularly given the current state of the unit testing). 

Aside from all that, my earlier attempt to retrofit tests into the watcher have left me convinced that adding tests to the watcher would require a non-trival refactor, and the metrics registry wouldn't be the deal breaker if/when that is undertaken.

Though I haven't dug into the agent's metrics since the code review, It seems like the registry really ought to be passed into the agent state constructor rather than the other way around.

Reply:
> It seems like the registry really ought to be passed into the agent state constructor rather than the other way around.

Yup, agree. Don't want to do that in this PR (or block on a separate PR to do that), but 100% should do that after merging.